django/django
# Fixed #28385 -- Avoided ignoring natural keys when primary key has a default value.

## Ticket

https://code.djangoproject.com/ticket/28385

## Bug space

When a models primary key has default set, and a fixture is loaded that does not have a pk attribute, then the serialisation code will refuse to update the existing instance and instead try to create a new record in the db.

When the primary key attribute has a default value - such as models using UUID primary keys - the model will fill the pk attribute with that at instantiation time, and the check in build_instance will see that default value and refuse to goto the database to fetch the existing instance.

This is against the expectations set in the docs, that when the model/manager implement the natural keys interfaces and the primary key is omitted, the framework will update the existing instance when deserialising.

(c) reporter Daniel Knell

## How to reproduce

1. Create the same model like [there](https://github.com/django/django/pull/8801/files#diff-2cf99154fd579c9e2d2ee0945f90ef4bR26).

2. Make `loaddata` fixtures flow like [there](https://github.com/django/django/pull/8801/files#diff-79e3f82c439838acf1377e2ec7e1342bR79).

We expect, that if `loaddata` dictionary has no `pk`, it will get object by `natural key` and update object, but it raises:

```
django.db.utils.IntegrityError: Problem installing fixture 'filename.json': 
Could not load app.Model(pk=<pk_number>): UNIQUE constraint failed: app_model.some_field
```
## Why it happens 

1. [Deserializer](https://github.com/django/django/blob/master/django/core/serializers/python.py#L98) flow does not provide `pk` for objects without it.
2. `Deserializer` try to get `pk` from `natural_key` in [build_instance](https://github.com/django/django/blob/master/django/core/serializers/base.py#L215) method.
3. But this method create new model object with default `pk` before implement point `2` [there](https://github.com/django/django/blob/master/django/core/serializers/base.py#L222).

Again: when model is created, we can get `pk` of this model (because `pk` has default value).
Question: how we can get model's `pk` by `natural key` from model without `pk`, if we create one instance before with default `pk`?

## Solution

I got `build_instance` logic and move it to previous level within [Deserializer](https://github.com/DmytryiStriletskyi/django/blob/8d2564a2805f2ebe48a9e5194b5c678f016f4e59/django/core/serializers/python.py#L79) to suitable place.

If it seems okay, we can go ahead and rewrite logic for [that](https://github.com/django/django/blob/d0f59054d0c4b2291a4e0e94ec00537f98a4ab50/django/core/serializers/xml_serializer.py#L213). Also, we can change `build_instance` logic for this two cases instead of touch serializers flow. 

It was not possible to use `TDD`. May add tests in last commit.
Thank you for paying attention.
